### PR TITLE
feat(ctb): `DisputeGameFactory` recent game search

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Adds a new convinence function to the `DisputeGameFactory` that makes it easier for off-chain actors to query the most recent `n` dispute games of a certain game type. For packages like `op-viem`, this is useful for determining a recent output proposal to prove a withdrawal against.

### CODE REVIEW

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* `develop`
  * **PR #9258** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/9258?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #9259** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/9259?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/9258?utm_source=stack-comment).

**Review by refcell — APPROVED** [review]
Looks good to me

File: packages/contracts-bedrock/src/dispute/DisputeGameFactory.sol [hunk 1]
```diff
@@ -133,6 +133,49 @@ contract DisputeGameFactory is OwnableUpgradeable, IDisputeGameFactory, ISemver
...
+        if (_start >= _disputeGameList.length || _n == 0) return games_;
+
+        // Allocate enough memory for the full array, but start the array's length at `0`. We may not use all of the
+        // memory allocated, but we don't know ahead of time the final size of the array.
+        assembly {
+            games_ := mload(0x40)
+            mstore(0x40, add(games_, add(0x20, shl(0x05, _n))))
+        }
+
+        // Perform a reverse linear search for the `_n` most recent games of type `_gameType`.
```
Comment (Inphi) [inline packages/contracts-bedrock/src/dispute/DisputeGameFactory.sol:156]: hmm, why do this in reverse if we want games added since `_start`?

Comment (clabby) [reply packages/contracts-bedrock/src/dispute/DisputeGameFactory.sol:156]: We want the most _recent_ games, and the dispute game list is ordered by time inserted. The reverse search starts from `_start` and then looks backwards for games added at or before `_start` 👍

Comment (refcell) [reply packages/contracts-bedrock/src/dispute/DisputeGameFactory.sol:156]: Resolving this so we can resume bond progress.

**Pull request merged**